control-plane-operator/controllers/hostedcontrolplane/cloud/kubevirt/reconcile.go

```diff
@@ -0,0 +1,135 @@
+package kubevirt
+
+import (
+	"fmt"
+
+	hyperv1 "github.com/openshift/hypershift/api/v1alpha1"
+	"github.com/openshift/hypershift/support/config"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func ccmServiceAccount(ns string) *corev1.ServiceAccount {
+	return &corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "kubevirt-cloud-controller-manager",
+			Namespace: ns,
+		},
+	}
+}
+
+func ccmRole(ns string) *rbacv1.Role {
+	return &rbacv1.Role{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "kubevirt-cloud-controller-manager",
+			Namespace: ns,
+		},
+	}
+}
+
+func ccmRoleBinding(ns string) *rbacv1.RoleBinding {
+	return &rbacv1.RoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "kubevirt-cloud-controller-manager",
+			Namespace: ns,
+		},
+	}
+}
+
+// func ccmClusterRole() *rbacv1.ClusterRole {
+// 	return &rbacv1.ClusterRole{
+// 		ObjectMeta: metav1.ObjectMeta{
+// 			Name: "kubevirt-cloud-controller-manager-clusterrole",
+// 		},
+// 	}
+// }
+//
+// func ccmClusterRoleBinding() *rbacv1.ClusterRoleBinding {
+// 	return &rbacv1.ClusterRoleBinding{
+// 		ObjectMeta: metav1.ObjectMeta{
+// 			Name: "kubevirt-cloud-controller-manager-clusterrole",
+// 		},
+// 	}
+// }
+
+func ReconcileCloudConfig(cm *corev1.ConfigMap, hcp *hyperv1.HostedControlPlane, kubeconfig []byte) error {
+	cfg := cloudConfig(kubeconfig)
+	serializedCfg, err := cfg.String()
+	if err != nil {
+		return fmt.Errorf("failed to serialize cloudconfig: %w", err)
+	}
+
+	if cm.Data == nil {
+		cm.Data = map[string]string{}
+	}
+	cm.Data[CloudConfigKey] = string(serializedCfg)
+
+	return nil
+}
+
+func reconcileCCMServiceAccount(sa *corev1.ServiceAccount, ownerRef config.OwnerRef) error {
+	ownerRef.ApplyTo(sa)
+	return nil
+}
+
+func reconcileCCMRole(role *rbacv1.Role, ownerRef config.OwnerRef) error {
+	ownerRef.ApplyTo(role)
+	role.Rules = []rbacv1.PolicyRule{
+		{
+			APIGroups: []string{"kubevirt.io"},
+			Resources: []string{"virtualmachines", "virtualmachineinstances"},
+			Verbs:     []string{rbacv1.VerbAll},
+		},
+		{
+			APIGroups: []string{""},
+			Resources: []string{"services", "pods"},
+			Verbs:     []string{rbacv1.VerbAll},
```

Comment:
the cloud provider controller doesn't need the ability to create/delete/update/patch pods, right?

Reply:
It need also to "update" the pod
When changed to "get", "list", "watch" and "update" permissions, got the following error:
```
loadbalancer.go:269] Failed to update pod virt-launcher-nargaman-hosted-djxbs-hrt2h: pods "virt-launcher-nargaman-hosted-djxbs-hrt2h" is forbidden: unable to validate against any security context constraint: [provider "anyuid": Forbidden: not usable by user or serviceaccount, provider "containerized-data-importer": Forbidden: not usable by user or serviceaccount, provider restricted-v2: .spec.securityContext.seLinuxOptions.level: Invalid value: "": must be s0:c30,c10, provider restricted-v2: .spec.securityContext.seLinuxOptions.type: Invalid value: "virt_launcher.process": must be , spec.containers[0].securityContext.runAsUser: Invalid value: 0: must be in the ranges: [1000890000, 1000899999], spec.containers[0].securityContext.seLinuxOptions.level: Invalid value: "": must be s0:c30,c10, spec.containers[0].securityContext.seLinuxOptions.type: Invalid value: "virt_launcher.process": must be , spec.containers[0].securityContext.capabilities.add: Invalid value: "SYS_NICE": capability may not be added, provider "restricted": Forbidden: not usable by user or serviceaccount, provider "nonroot-v2": Forbidden: not usable by user or serviceaccount, provider "nonroot": Forbidden: not usable by user or serviceaccount, provider "hostmount-anyuid": Forbidden: not usable by user or serviceaccount, provider "kubevirt-controller": Forbidden: not usable by user or serviceaccount, provider "bridge-marker": Forbidden: not usable by user or serviceaccount, provider "machine-api-termination-handler": Forbidden: not usable by user or serviceaccount, provider "hostnetwork-v2": Forbidden: not usable by user or serviceaccount, provider "hostnetwork": Forbidden: not usable by user or serviceaccount, provider "hostaccess": Forbidden: not usable by user or serviceaccount, provider "linux-bridge": Forbidden: not usable by user or serviceaccount, provider "nmstate": Forbidden: not usable by user or serviceaccount, provider "kubevirt-handler": Forbidden: not usable by user or serviceaccount, provider "node-exporter": Forbidden: not usable by user or serviceaccount, provider "privileged": Forbidden: not usable by user or serviceaccount]
```
I'm checking this